Commit: show summary when commiting

yyyc514 opened the pull request:
Relates to #107.

yyyc514 commented:
If this is heading the right direction I can improve the syntax further.

yyyc514 commented:
I bumped the help to `###` cause I thought it looked better.

divmain commented on core/commands/commit.py:
```diff
@@ -57,11 +59,37 @@ class GsCommitInitializeViewCommand(TextCommand, GitCommand):
 
-        option_amend = self.view.settings().get("git_savvy.commit_view.amend")
-        if option_amend:
+        if self.amending():
```
This is an unnecessary change.

divmain commented on core/commands/commit.py:
```diff
@@ -71,11 +99,14 @@ def run(self, edit):
                 initial_text += self.git("diff", "--cached")
 
+        if sublime.load_settings("GitSavvy.sublime-settings").get("show_commit_summary"):
+            initial_text += self.fetch_commit_summary()
```
Just pass in `option_amend` here.

yyyc514 replied:
Is there some reason we can't wrap it in a method vs passing it everywhere?  `amending` really does seem to be a property of `GsCommitInitializeViewCommand to me`.

yyyc514 replied:
My next idea was to clean this up so it's just content = " ".[header(), help_text(), footer()].  Since lots of different pieces of the render need to know about amend it seems logical to make it accessible on the class vs passing it into every render method.

divmain commented on core/commands/commit.py:
```diff
@@ -57,11 +59,37 @@ class GsCommitInitializeViewCommand(TextCommand, GitCommand):
+        dryrun = self.git("commit", "--dry-run", *args, throw_on_stderr = False)
+        # detect trying to do a commit with nothing staged
+        if dryrun.find('no changes added') != -1:
+            sublime.error_message(NO_STAGED_FILES)
```
If you go this route, there should be an exit mechanism - after showing this error message, it'll continue one with executing this method.

yyyc514 replied:
I wasn't sure what the proper thing to do was since at that point we already have the view (it's created by the first command at the window level).

divmain commented:
I think a more straightforward approach here would be something along the lines of the following.

```
status = self.git("status")
status_comment = "\n".join("## " + line for line in status.split("\n"))
```

Really, the normal behavior here for commits is to display the verbose `git status` output.

divmain commented on core/commands/commit.py:
```diff
@@ -5,16 +5,18 @@
+### You may also reference or close a GitHub issue with this commit.  To do so,
+### type `#` followed by the `tab` key.  You will be shown a list of issues
+### related to the current repo.  You may also type `owner/repo#` plus the `tab`
+### key to reference an issue in a different GitHub repo.
```
I'd prefer to keep it at two `#`.

yyyc514 commented:
`status = self.git("status")`

There is no --amend option for status.  Since what we want is to know exactly what the commit is planning to do there is no substitute for asking the commit command itself.

divmain commented:
Hmm that's a very good point @yyyc514 .  Let me think on this a bit longer.  Sorry if it feels like I took an axe to this - just got off a plane and pretty tired.  Very much appreciate the PR.  As much as possible I prefer to keep things a simple as possible until a clean abstraction becomes apparent, which is where most of my comments are coming from.  I'll re-review this tomorrow.

divmain commented on syntax/make_commit.YAML-tmLanguage:
```diff
@@ -6,7 +6,15 @@ uuid: 0d4c19fb-b197-4d90-806e-7212e542c86b
+  begin: ^#[^#]
+  end: $
+  patterns:
+  - comment: git gile categories
```
`git file categories`?

yyyc514 replied:
Open to suggestions of a better name :)

divmain replied:
I was just pointing out the typo :)

yyyc514 replied:
And I still didn't see it. LOL.  Ok.

divmain commented:
Alright, I'm going to hit the sack, but I'll take a look at this again tomorrow.  Thanks again.  One last nit-pick before I sign-off - the top method in each class should be `run`, for consistency.

yyyc514 commented:
> One last nit-pick before I sign-off - the top method in each class should be run, for consistency.

Oh sure, that's no problem.

yyyc514 commented:
I'll be more than happy to port this stuff to the new Interface framework when that is all polished up.

yyyc514 commented:
Rebased and squashed.

divmain commented:
I'm traveling for the next few days and will want some time to dig into this before merging.  This PR will have to wait for a couple of days.

yyyc514 commented:
I've rebased and removed the error case.  Currently commit is not an error case if there are no files staged, so this probably shouldn't be either.  Plus if there are no files staged it's pretty easy to notice that when you're showing the dry-run.

Ping.

divmain commented:
Not ignoring this one.  I'm mulling over how I want to approach the commit view, after working with the other interfaces.  Leaving this PR open and, even if it is not merged, I'll reference what you did here.

stoivo commented:
Hey, how is the planing going

divmain commented:
Closing this, but issue will be addressed as part of #50.
